src/main/java/org/ossgang/commons/monads/AsyncMaybe.java

```diff
@@ -0,0 +1,284 @@
+// @formatter:off
+/*******************************************************************************
+ *
+ * This file is part of ossgang-commons.
+ *
+ * Copyright (c) 2008-2019, CERN. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+// @formatter:on
+
+package org.ossgang.commons.monads;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+import static org.ossgang.commons.utils.Exceptions.unchecked;
+
+/**
+ * TODO andrea
+ *
+ * @param <T>
+ */
+public class AsyncMaybe<T> {
+
+    private final CompletableFuture<T> future;
+    private final Function<Supplier<T>, Maybe<T>> maybeGenerator;
+
+    protected AsyncMaybe(CompletableFuture<T> future, Function<Supplier<T>, Maybe<T>> maybeGenerator) {
+        this.future = future;
+        this.maybeGenerator = maybeGenerator;
+    }
+
+    public static <T> AsyncMaybe<T> fromCompletableFuture(CompletableFuture<T> future) {
+        return new AsyncMaybe<>(future, valueSupplier -> Maybe.attempt(valueSupplier::get));
+    }
+
+    public static AsyncMaybe<Void> fromVoidCompletableFuture(CompletableFuture<Void> future) {
+        return new AsyncMaybe<>(future, valueSupplier -> {
+            try {
+                valueSupplier.get();
```

Comment:
Is there a generics problem here?
In principle there should be a Maybe.attempt(Runnable) which returns a Maybe of Void.
So you should be able to just do ```valueSupplier -> Maybe.attempt(() -> valueSupplier.get())```

Reply:
Well spotted :smile: